cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -196,100 +234,127 @@ public final ProgramController run(final Program program, final ProgramOptions o
+                .withBundlerClassAcceptor(new HadoopClassExcluder() {
+                  @Override
+                  public boolean accept(String className, URL classUrl, URL classPathUrl) {
+                    // Exclude both hadoop and spark classes.
+                    return super.accept(className, classUrl, classPathUrl)
+                      && !className.startsWith("org.apache.spark.");
+                  }
+                })
+                .withApplicationArguments(
+                  "--" + RunnableOptions.JAR, programJarName,
+                  "--" + RunnableOptions.HADOOP_CONF_FILE, HADOOP_CONF_FILE_NAME,
+                  "--" + RunnableOptions.CDAP_CONF_FILE, CDAP_CONF_FILE_NAME,
+                  "--" + RunnableOptions.APP_SPEC_FILE, APP_SPEC_FILE_NAME,
+                  "--" + RunnableOptions.PROGRAM_OPTIONS, programOptions,
+                  "--" + RunnableOptions.PROGRAM_ID, GSON.toJson(program.getId().toEntityId())
+                );
+
+              TwillController twillController;
+              // Change the context classloader to the combine classloader of this ProgramRunner and
+              // all the classloaders of the dependencies classes so that Twill can trace classes.
+              ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(new CombineClassLoader(
+                AbstractDistributedProgramRunner.this.getClass().getClassLoader(),
+                Iterables.transform(dependencies, new Function<Class<?>, ClassLoader>() {
+                  @Override
+                  public ClassLoader apply(Class<?> input) {
+                    return input.getClassLoader();
+                  }
+                })));
+              try {
+                twillController = twillPreparer.start();
+              } finally {
+                ClassLoaders.setContextClassLoader(oldClassLoader);
+              }
+              return addCleanupListener(twillController, program, tempDir);
+            }
+          });
         }
-      });
+      };
+
+      if (SecurityUtil.isKerberosEnabled(cConf)) {
+        String principal = options.getArguments().getOption(ProgramOptionConstants.PRINCIPAL);
+        String keytabPath = options.getArguments().getOption(ProgramOptionConstants.KEYTAB_PATH);
+        LOG.info("Configured principal={}, keytab={}", principal, keytabPath);
+
+        String expandedPrincipal = SecurityUtil.expandPrincipal(principal);
+        String localizedKeytabPath = getKeytabPath(URI.create(keytabPath), tempDir);
+        LOG.info("Logging in as: principal={}, keytab={}", expandedPrincipal, localizedKeytabPath);
+
+        UserGroupInformation aliUGI =
+          UserGroupInformation.loginUserFromKeytabAndReturnUGI(expandedPrincipal, localizedKeytabPath);
+        return ProgramRunners.runAsUGI(aliUGI, callable);
+      } else {
+        // otherwise, simply launch as the current user
+        return callable.call();
```

Comment:
nit: reverse the condition on line 341 to just have this in the if, so you reduce the indentation?